File: rest-service/src/main/java/mb/api/model/blast/IOBlastxConfig.java
package mb.api.model.blast;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import mb.api.model.blast.impl.IOBlastxConfigImpl;
import mb.lib.blast.model.CompBasedStats;
import org.veupathdb.lib.blast.BlastTool;
import org.veupathdb.lib.blast.field.BlastXTask;
import org.veupathdb.lib.blast.field.ScoringMatrix;
import org.veupathdb.lib.blast.field.Seg;
import org.veupathdb.lib.blast.field.Strand;

@JsonTypeName("blastx")
@JsonDeserialize(as = IOBlastxConfigImpl.class)
public interface IOBlastxConfig extends IOBlastConfig
{
  BlastTool _DISCRIMINATOR_TYPE_NAME = BlastTool.BlastX;

  @JsonProperty("strand")
  Strand getStrand();

  @JsonProperty("strand")
  void setStrand(Strand strand);

  @JsonProperty("queryGeneticCode")
  Short getQueryGeneticCode();

  @JsonProperty("queryGeneticCode")
  void setQueryGeneticCode (Short queryGeneticCode);

  @JsonProperty("task")
  BlastXTask getTask();

  @JsonProperty("task")
  void setTask(BlastXTask task);

  @JsonProperty("wordSize")
  Integer getWordSize();

  @JsonProperty("wordSize")
  void setWordSize(Integer wordSize);

  @JsonProperty("gapOpen")
  Integer getGapOpen();

  @JsonProperty("gapOpen")
  void setGapOpen(Integer gapOpen);

  @JsonProperty("gapExtend")
  Integer getGapExtend();

  @JsonProperty("gapExtend")
  void setGapExtend(Integer gapExtend);

  @JsonProperty("maxIntronLength")
  Integer getMaxIntronLength();

  @JsonProperty("maxIntronLength")
  void setMaxIntronLength(Integer maxIntronLength);

  @JsonProperty("matrix")
  ScoringMatrix getMatrix();

  @JsonProperty("matrix")
  void setMatrix(ScoringMatrix matrix);

  @JsonProperty("threshold")
   Double getThreshold();

  @JsonProperty("threshold")
  void setThreshold (Double threshold);

  @JsonProperty("compBasedStats")
  CompBasedStats getCompBasedStats();

  @JsonProperty("compBasedStats")
  void setCompBasedStats(CompBasedStats compBasedStats);

  @JsonProperty("seg")
  Seg getSeg();

  @JsonProperty("seg")
  void setSeg(Seg seg);

  @JsonProperty("softMasking")
   Boolean getSoftMasking();

  @JsonProperty("softMasking")
  void setSoftMasking (Boolean softMasking);

  @JsonProperty("taxIds")
  List<Integer> getTaxIds();

  @JsonProperty("taxIds")
  void setTaxIds(List<Integer> taxIds);

  @JsonProperty("negativeTaxIds")
  List<Integer> getNegativeTaxIds();

  @JsonProperty("negativeTaxIds")
  void setNegativeTaxIds(List<Integer> negativeTaxIds);

  @JsonProperty("dbSoftMask")
  String getDbSoftMask();

  @JsonProperty("dbSoftMask")
  void setDbSoftMask(String dbSoftMask);

  @JsonProperty("dbHardMask")
  String getDbHardMask();

  @JsonProperty("dbHardMask")
  void setDbHardMask(String dbHardMask);

  @JsonProperty("cullingLimit")
  Integer getCullingLimit();

  @JsonProperty("cullingLimit")
  void setCullingLimit(Integer cullingLimit);

  @JsonProperty("bestHitOverhang")
   Double getBestHitOverhang();

  @JsonProperty("bestHitOverhang")
  void setBestHitOverhang (Double bestHitOverhang);

  @JsonProperty("bestHitScoreEdge")
   Double getBestHitScoreEdge();

  @JsonProperty("bestHitScoreEdge")
  void setBestHitScoreEdge (Double bestHitScoreEdge);

  @JsonProperty("subjectBestHit")
   Boolean getSubjectBestHit();

  @JsonProperty("subjectBestHit")
  void setSubjectBestHit (Boolean subjectBestHit);

  @JsonProperty("sumStats")
   Boolean getSumStats();

  @JsonProperty("sumStats")
  void setSumStats (Boolean sumStats);

  @JsonProperty("xDropGap")
   Double getXDropGap();

  @JsonProperty("xDropGap")
  void setXDropGap (Double xDropGap);

  @JsonProperty("xDropGapFinal")
   Double getXDropGapFinal();

  @JsonProperty("xDropGapFinal")
  void setXDropGapFinal (Double xDropGapFinal);

  @JsonProperty("windowSize")
  Integer getWindowSize();

  @JsonProperty("windowSize")
  void setWindowSize(Integer windowSize);

  @JsonProperty("ungapped")
   Boolean getUngapped();

  @JsonProperty("ungapped")
  void setUngapped (Boolean ungapped);

  @JsonProperty("useSWTraceback")
   Boolean getUseSWTraceback();

  @JsonProperty("useSWTraceback")
  void setUseSWTraceback (Boolean useSWTraceback);
}

File: rest-service/src/main/java/mb/api/model/blast/IOBlastxConfig.kt
package mb.api.model.blast

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_DEFAULT
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeName
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import mb.api.model.blast.impl.IOBlastxConfigImpl
import mb.lib.blast.model.CompositionBasedStats
import org.veupathdb.lib.blast.BlastTool
import org.veupathdb.lib.blast.field.BlastXTask
import org.veupathdb.lib.blast.field.ScoringMatrix
import org.veupathdb.lib.blast.field.Seg
import org.veupathdb.lib.blast.field.Strand

@JsonTypeName("blastx")
@JsonDeserialize(`as` = IOBlastxConfigImpl::class)
@JsonInclude(NON_DEFAULT)
interface IOBlastxConfig : IOBlastWithLists {
  @get:JsonProperty("strand")
  @set:JsonProperty("strand")
  var strand: Strand?

  @get:JsonProperty("queryGeneticCode")
  @set:JsonProperty("queryGeneticCode")
  var queryGeneticCode: Short?

  @get:JsonProperty("task")
  @set:JsonProperty("task")
  var task: BlastXTask?

  @get:JsonProperty("wordSize")
  @set:JsonProperty("wordSize")
  var wordSize: Int?

  @get:JsonProperty("gapOpen")
  @set:JsonProperty("gapOpen")
  var gapOpen: Int?

  @get:JsonProperty("gapExtend")
  @set:JsonProperty("gapExtend")
  var gapExtend: Int?

  @get:JsonProperty("maxIntronLength")
  @set:JsonProperty("maxIntronLength")
  var maxIntronLength: Int?

  @get:JsonProperty("matrix")
  @set:JsonProperty("matrix")
  var matrix: ScoringMatrix?

  @get:JsonProperty("threshold")
  @set:JsonProperty("threshold")
  var threshold: Double?

  @get:JsonProperty("compBasedStats")
  @set:JsonProperty("compBasedStats")
  var compBasedStats: CompositionBasedStats?

  @get:JsonProperty("seg")
  @set:JsonProperty("seg")
  var seg: Seg?

  @get:JsonProperty("softMasking")
  @set:JsonProperty("softMasking")
  var softMasking: Boolean

  @get:JsonProperty("dbSoftMask")
  @set:JsonProperty("dbSoftMask")
  var dbSoftMask: String?

  @get:JsonProperty("dbHardMask")
  @set:JsonProperty("dbHardMask")
  var dbHardMask: String?

  @get:JsonProperty("cullingLimit")
  @set:JsonProperty("cullingLimit")
  var cullingLimit: Int?

  @get:JsonProperty("bestHitOverhang")
  @set:JsonProperty("bestHitOverhang")
  var bestHitOverhang: Double?

  @get:JsonProperty("bestHitScoreEdge")
  @set:JsonProperty("bestHitScoreEdge")
  var bestHitScoreEdge: Double?

  @get:JsonProperty("subjectBestHit")
  @set:JsonProperty("subjectBestHit")
  var subjectBestHit: Boolean

  @get:JsonProperty("sumStats")
  @set:JsonProperty("sumStats")
  var sumStats: Boolean?

  @get:JsonProperty("xDropGap")
  @set:JsonProperty("xDropGap")
  var xDropGap: Double

  @get:JsonProperty("xDropGapFinal")
  @set:JsonProperty("xDropGapFinal")
  var xDropGapFinal: Double

  @get:JsonProperty("windowSize")
  @set:JsonProperty("windowSize")
  var windowSize: Int?

  @get:JsonProperty("ungapped")
  @set:JsonProperty("ungapped")
  var ungapped: Boolean?

  @get:JsonProperty("useSWTraceback")
  @set:JsonProperty("useSWTraceback")
  var useSWTraceback: Boolean?

  companion object {
    val _DISCRIMINATOR_TYPE_NAME = BlastTool.BlastX
  }
}